### PR TITLE
Fix unit tests when run on MS-Windows

### PR DESCRIPTION
**Pull request by bbrodt**
See https://issues.jboss.org/browse/UF-463

### CODE REVIEW

File: uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-backend/src/test/java/org/uberfire/ext/plugin/backend/PluginSamples.java [hunk 1]
```diff
@@ -97,7 +97,7 @@ public static PluginSimpleContent getScreenWithMainAndTitle() {
     public static String loadSample( String file ) {
         try {
             return IOUtils.toString( new PluginSamples().getClass().getResourceAsStream( file ),
-                                     "UTF-8" );
+                                     "UTF-8" ).replace( "\r\n", "\n" );
```
Comment (ederign) [inline uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-backend/src/test/java/org/uberfire/ext/plugin/backend/PluginSamples.java]: Why don't we move this function to https://github.com/uberfire/uberfire/tree/master/uberfire-testing-utils and avoid the copy and paste?

File: uberfire-io/src/main/java/org/uberfire/io/impl/AbstractIOService.java [hunk 2]
```diff
@@ -636,10 +636,16 @@ public Path write( final Path path,
 
         try {
             byteChannel.write( ByteBuffer.wrap( bytes ) );
-            byteChannel.close();
         } catch ( final java.io.IOException e ) {
             throw new IOException( e );
         }
+        finally {
+            try {
```
Comment (ederign) [inline uberfire-io/src/main/java/org/uberfire/io/impl/AbstractIOService.java]: I don't like nested try catch blocks and also silently swallowing an error condition and continuing execution.

File: uberfire-nio2-backport/uberfire-nio2-api/src/main/java/org/uberfire/java/nio/file/Paths.java [hunk 3]
```diff
@@ -46,6 +47,10 @@ public static Path get(final String first, final String... more)
 
         URI uri = null;
         if (more == null || more.length == 0) {
+            if (AbstractPath.hasWindowsDriver( first )) {
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-api/src/main/java/org/uberfire/java/nio/file/Paths.java]: Tests?

File: uberfire-nio2-backport/uberfire-nio2-api/src/test/java/org/uberfire/java/nio/file/PathsTest.java [hunk 4]
```diff
@@ -60,7 +60,7 @@ public void simplePartitionedWindowsGet() {
         assertThat( path ).isNotNull();
         assertThat( path.isAbsolute() ).isTrue();
 
-        assertThat( path.toString() ).isEqualTo( "c:\\path" + separator() + "to" + separator() + "file.txt" );
+        assertThat( path.toString() ).isEqualTo( "c:\\path" + "\\" + "to" + "\\" + "file.txt" );
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-api/src/test/java/org/uberfire/java/nio/file/PathsTest.java]: why did you remove separator() and hard code "\\"?

File: uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/main/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStore.java [hunk 5]
```diff
@@ -52,10 +53,21 @@ public SimpleWindowsFileStore( final File[] roots,
         this.roots = roots;
 
         if ( path.isAbsolute() ) {
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/main/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStore.java]: Please add tests for this.

File: uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/main/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStore.java [hunk 6]
```diff
@@ -73,7 +85,9 @@ public SimpleWindowsFileStore( final File[] roots,
 
     @Override
     public String name() {
-        return listRoots()[ fstoreIndex ].getName();
+        // the File.getName() for the root path "C:/" will always be an empty string
+        // so use File.getPath() instead
+        return listRoots()[ fstoreIndex ].getPath();
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/main/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStore.java]: On *nix listRoots()[ fstoreIndex ].getName() == listRoots()[ fstoreIndex ].getPath() ?

File: uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/test/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStoreTest.java [hunk 7]
```diff
@@ -42,17 +43,31 @@
     final File[]             roots       = new File[]{ new File( "c:\\" ), new File( "a:\\" ) };
     final FileSystem         fileSystem  = new SimpleWindowsFileSystem( roots, fsProvider, "c:\\" );
     final Path               nonNullPath = GeneralPathImpl.create( fileSystem, "c:\\something", false );
+    boolean isWindows;
+    
+    @Before
+    public void initialize() {
+        isWindows = System.getProperty( "os.name" ).toLowerCase().contains( "windows" );
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/test/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStoreTest.java]: What do you think about moving this copy and paste code to testing utils?

File: uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/AbstractPath.java [hunk 8]
```diff
@@ -48,8 +49,15 @@
         implements Path,
                    AttrHolder {
 
-    public static final Pattern WINDOWS_DRIVER = Pattern.compile( "^/?[A-Z|a-z]+(:).*" );
+    // NOTE: the Windows drive letter parsing was incorrect: it allowed more
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/AbstractPath.java]: If you fix the issue, you should probably remove this comment.

File: uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/AbstractPath.java [hunk 9]
```diff
@@ -95,19 +103,25 @@ protected AbstractPath( final FS fs,
...
-        final RootInfo rootInfo = setupRoot( fs, path, host, isRoot );
-        this.path = rootInfo.path;
+        this.usesWindowsFormat = path.contains( WINDOWS_SEPARATOR_STRING );
 
+        String pathx = null;
+        if ( path.startsWith( UNIX_SEPARATOR_STRING ) && hasWindowsDriver( path ) ) {
+               pathx = path.substring(1);
+        }
+        else
+            pathx = path;
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/AbstractPath.java]: please always use {} in your else blocks.

File: uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/AbstractPath.java [hunk 10]
```diff
@@ -588,13 +608,62 @@ public String toString( boolean addHost ) {
...
     }
 
+    public static String appendTrailingSeparator( final String path ) {
+        if ( path.length() > 1)
+            return removeTrailingSeparator( path ) + getSeparator( path );
+        return path;
+    }
+    
+    public static boolean hasWindowsDriver( final String path ) {
+//        checkNotEmpty( "path", path );
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/AbstractPath.java]: ?

File: uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/AbstractPath.java [hunk 11]
```diff
@@ -632,14 +701,34 @@ public boolean equals( final Object o ) {
             return false;
         }
 
-        if ( !usesWindowsFormat && !Arrays.equals( path, other.path ) ) {
+//        if ( !usesWindowsFormat && !Arrays.equals( path, other.path ) ) {
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/AbstractPath.java]: dead code?

File: uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/AbstractPath.java [hunk 12]
```diff
@@ -48,8 +49,15 @@
         implements Path,
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/AbstractPath.java]: You should add tests for every piece of code that changes the behaviour. This is a really sensitive area in Uberfire.

File: uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/GeneralPathImpl.java [hunk 13]
```diff
@@ -79,19 +79,14 @@ protected RootInfo setupRoot( final FileSystem fs,
     @Override
     protected String defaultDirectory() {
         if ( usesWindowsFormat ) {
-            final String result = new File( "" ).getAbsolutePath().replaceAll( "/", "\\\\" ) + "\\";
+            final String result = new File( "" ).getAbsolutePath().replace( UNIX_SEPARATOR_STRING, WINDOWS_SEPARATOR_STRING ) + WINDOWS_SEPARATOR_STRING;
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/GeneralPathImpl.java]: tests?

File: uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/dotfiles/DotFileUtils.java [hunk 14]
```diff
@@ -36,11 +38,46 @@ private DotFileUtils() {
...
         boolean hasContent = false;
-        if (attrs != null && attrs.length > 0) {
+        try {
+            hasContent = buildDotFile( path, out, attrs);
+        }
+        finally {
+            try {
+                out.close();
+            } catch (Exception e) {
+            }
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/dotfiles/DotFileUtils.java]: Empty catch block

File: uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/dotfiles/DotFileUtils.java [hunk 15]
```diff
@@ -36,11 +38,46 @@ private DotFileUtils() {
...
+            out = new FileOutputStream( dotFile );
+            hasContent = buildDotFile( path, out, attrs);
+        }
+        catch (Exception e) {
+        }
+        finally {
+            try {
+                out.close();
+            } catch (Exception e) {
+            }
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/dotfiles/DotFileUtils.java]: Empty catch block, nested try catch

File: uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/dotfiles/DotFileUtils.java [hunk 16]
```diff
@@ -63,13 +100,19 @@ public static boolean buildDotFile(final Path path,
...
+                }
+            }
+            path.getFileSystem().provider().deleteIfExists( dot( path ) );
         }
 
         if (!hasContent) {
             try {
                 out.close();
-            } catch (java.io.IOException e) {
+            } catch ( Exception e ) {
```
Comment (ederign) [inline uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/dotfiles/DotFileUtils.java]: empty catch block

File: uberfire-server/src/test/java/org/uberfire/server/locale/GWTLocaleHeaderFilterTest.java [hunk 17]
```diff
@@ -50,7 +50,9 @@ public void testLocaleDefault() throws IOException, ServletException {
 
         localeHeaderFilter.doFilter( req, resp, chain );
 
-        assertEquals( new Scanner( getClass().getResourceAsStream( "/expected-sample.html" ), "UTF-8" ).useDelimiter( "\\A" ).next(), sw.toString() );
+        String s1 = new Scanner( getClass().getResourceAsStream( "/expected-sample.html" ), "UTF-8" ).useDelimiter( "\\A" ).next().replace( "\r\n", "\n" );
```
Comment (ederign) [inline uberfire-server/src/test/java/org/uberfire/server/locale/GWTLocaleHeaderFilterTest.java]: Again the same pattern "\r\n", "\n" (you should move this to testing utils)

File: uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-backend/src/test/java/org/uberfire/ext/plugin/backend/PluginSamples.java (hunk 1 above)
Comment (bbrodt) [inline uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-backend/src/test/java/org/uberfire/ext/plugin/backend/PluginSamples.java]: I don't understand what you mean by copy and paste. Where is this duplicated?

File: uberfire-io/src/main/java/org/uberfire/io/impl/AbstractIOService.java (hunk 2 above)
Comment (bbrodt) [inline uberfire-io/src/main/java/org/uberfire/io/impl/AbstractIOService.java]: How would you rewrite this nested try-catch? What should I do if the bytechannel can't be closed?

File: uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-backend/src/test/java/org/uberfire/ext/plugin/backend/PluginSamples.java (hunk 1 above)
Comment (ederign) [inline uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-backend/src/test/java/org/uberfire/ext/plugin/backend/PluginSamples.java]: There is a lot of places with .replace( "\r\n", "\n" ).

File: uberfire-nio2-backport/uberfire-nio2-api/src/main/java/org/uberfire/java/nio/file/Paths.java (hunk 3 above)
Comment (bbrodt) [inline uberfire-nio2-backport/uberfire-nio2-api/src/main/java/org/uberfire/java/nio/file/Paths.java]: ok

File: uberfire-nio2-backport/uberfire-nio2-api/src/test/java/org/uberfire/java/nio/file/PathsTest.java (hunk 4 above)
Comment (bbrodt) [inline uberfire-nio2-backport/uberfire-nio2-api/src/test/java/org/uberfire/java/nio/file/PathsTest.java]: when this test is run on *nix, separator() returns "/", on windows it returns "\" so the two strings will never compare equal on windows. The whole effort behind the AbstractPath constructor change was to ensure consistent use of "/" and "\" within a Path - that's why some of these path tests failed on windows.

File: uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/main/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStore.java (hunk 5 above)
Comment (bbrodt) [inline uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/main/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStore.java]: okey dokey

File: uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/test/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStoreTest.java (hunk 7 above)
Comment (bbrodt) [inline uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/test/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStoreTest.java]: you got it dude 👍

File: uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/main/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStore.java (hunk 6 above)
Comment (bbrodt) [inline uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/main/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStore.java]: Yes

File: uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/GeneralPathImpl.java (hunk 13 above)
Comment (bbrodt) [inline uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/GeneralPathImpl.java]: Behavior hasn't changed, only replaced string constants with symbolic equivalents. I'm trying to eliminate the use of string constants for "/" and "\\" as much as possible to avoid future issues with assumptions about the path separator in use.

File: uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/main/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStore.java (hunk 6 above)
Comment (bbrodt) [inline uberfire-nio2-backport/uberfire-nio2-impls/uberfire-nio2-fs/src/main/java/org/uberfire/java/nio/fs/file/SimpleWindowsFileStore.java]: And BTW, Simple**Unix**FileStore.name() is hard-coded to return "/".

File: uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/AbstractPath.java (hunk 12 above)
Comment (bbrodt) [inline uberfire-nio2-backport/uberfire-nio2-model/src/main/java/org/uberfire/java/nio/base/AbstractPath.java]: OK, created additional AbstractPath tests in FileSystemsTest, GeneralPathEqualsTest, PathsTest and SimpleWindowsFileStoreTest

File: uberfire-io/src/main/java/org/uberfire/io/impl/AbstractIOService.java [hunk 18]
```diff
@@ -636,10 +636,16 @@ public Path write( final Path path,
 
         try {
             byteChannel.write( ByteBuffer.wrap( bytes ) );
-            byteChannel.close();
         } catch ( final java.io.IOException e ) {
             throw new IOException( e );
         }
+        if (byteChannel!=null) {
```
Comment (ederign) [inline uberfire-io/src/main/java/org/uberfire/io/impl/AbstractIOService.java]: why don't use a finally?

File: uberfire-io/src/test/java/org/uberfire/io/CommonIOExceptionsServiceDotFileTest.java [hunk 19]
```diff
@@ -65,7 +68,11 @@ public void newByteChannelFileAlreadyExistsException() {
 
         ioService().write( path, "ooooo!" );
 
-        ioService().newByteChannel( path );
+        ByteChannel channel = ioService().newByteChannel( path );
+        try {
+            channel.close();
+        } catch ( IOException e ) {
```
Comment (ederign) [inline uberfire-io/src/test/java/org/uberfire/io/CommonIOExceptionsServiceDotFileTest.java]: better throw the exception in the test to make it red.

**Comment by ederign** [comment]
@bbrodt thanks for your changes. Two minor comments.

Also, you have to rebase your PR and more important do a sanity check with linux + workbench in order to make sure that doesn't break on *nix environments.

@manstis could you please also take a look on this?

Thanks.

**Comment by psiroky** [comment]
Any news on this one? I would be great to have a green build also on Windows.

**Comment by bbrodt** [comment]
@ederign any news on this?

**Comment by bbrodt** [comment]
I can't merge this mess - too much water under the bridge. I'm closing this PR and creating a new one.
